stdio/file.c

```diff
@@ -581,37 +603,61 @@ char *fgets(char *str, int n, FILE *stream)
 
 int fflush_unlocked(FILE *stream)
 {
+	int ret = 0;
+
 	if (stream == NULL) {
-		fflush_unlocked(stdout);
-		return 0; /* TODO: flush all FILE's */
+		mutexLock(file_common.lock);
+		if (file_common.list != NULL) {
+			FILE *iter = file_common.list;
+			do {
+				mutexLock(iter->lock);
+				if ((iter->flags & F_WRITING) != 0) {
+					ret = fflush_unlocked(iter);
+				}
+				mutexUnlock(iter->lock);
+				if (ret != 0) {
+					break;
+				}
+				iter = iter->next;
+			} while (iter != file_common.list);
+		}
+		mutexUnlock(file_common.lock);
 	}
-
-	if (stream->flags & F_WRITING) {
-		if (stream->bufpos) {
-			full_write(stream->fd, stream->buffer, stream->bufpos);
-			stream->bufpos = 0;
+	else if ((stream->flags & F_WRITING) != 0) {
```

Comment:
maybe move to separate func `fflush_single`  or move stream == NULL semantics to `fflush` for better readability (and avoid recursion-at-a-glance)?

Reply:
I'd rather not create a new function, really, less better.

Hmmm, I placed this in unlocked to handle `fflush_unlocked(NULL`) too, but I guess I handle it wrong, as I get locks anyway. I wonder, should `fflush_unlocked(NULL)` work too?

Reply:
I don't see it stated explicitly, but I guess `fflush_unlocked(NULL)` should be handled

Reply:
"I'd rather not", but in the end I added not one but two functions, lol

Reply:
hehh, `fflush_unlocked` is linux-specific and implemented only in `glibc` so IMHO we don't need to focus on correctness here. I've checked the `glibc` implementation and `fflush_unlocked(NULL)` is the same as `fflush(NULL)` (acquires per-file locks - https://github.com/lattera/glibc/blob/master/libio/iofflush_u.c ). Surprisingly - there is an unlocked version which is used in cleanup (`fcloseall` implemetation and probably destructors).

I think You can either leave it as-is or remove unlocked NULL version for code simplicity.